### ISSUE TITLE
Initial feedback

### ISSUE THREAD

**Issue by Birkbjo** [dhis2/mar-y#47]
Really nice that we have this up and running now! These are my initial thought and open questions on the setup. I won't be too nitpicky, as the most important part for me is getting the data-handling part right - and at the moment that is kind of blocked by the Data-Engine.

## Structure

At first I was skeptical to the seperate redux-folder, but after some consideration I actually don't think we will have many features that have state in redux, so it might be nice to have this in one place, instead of the 'feature/duck'-way. Import paths for selectors might be a bit annoying though. 
I do however think that 'components' should be grouped into 'feature'-folders. They kind of are now, but I would move files like 'ProtectedRoute.js' and 'Navigation.js' into 'Navigation' folder.

- I would move each file in the root of the "components"-folders into a sub-folder, some of the folders currently only have .css, I would expect these to live with the .js files.

- I'm unsure about what should go in the 'pages'-folder. We are probably going to have generic components like 'ModelTypeList' and 'ModelForm'. Are these going in the page-folder or component-folder? Or is this just to setup routes for "overrides" (I know, that word has bad reputation by now).

## Other
- We need an 'initial'-loader. This needs to load userSettings, systemSettings, userAuthorities, userDataStore. This should probably be stored in redux. This also means that we can remove the queries for Navigation in `useHasAuthority`, as these can be checked synchronously from the redux-store after initialization.
   - Decide on how much schema-information we want to utilize, and if so; we may need to load these on initial-load as well.
- Right now the NavigationLinkComponent will send a request for every rendered link. This is mostly Data-Engines fault, but can be improved by the above point.
   - Ask backend for /me/authorization field-filtering. I still believe having one bigger request on load for this is better than using the Data-Engine to send requests for each link.
- I don't think we need 'connected-react-router'. Use [history](https://reacttraining.com/react-router/web/api/Router/history-object) and push from there if we need to do this imperatively. We should use Link-components instead of onClick-handlers calling 'push', unless there's a good reason not to. 
- I think sectionOrder can be simplified by moving the options-object to the section-objects. I was initially confused by the optional array-with-option API.

## Way forward
We need to see how Data-Engine evolves and how we would treat data. I will put this in another issue, to show complex examples we need to be able to handle. Keywords: normalization, 'view-data-transformation' (might be done with selectors?). 

- Generic selectors for modelTypes. We can't write selectors for all of the objects, we need some kind of generic way to handle this
- Forms integration, this is a crucial part of the app, and I'm excited to experiment with this. 
  - Did we decide on a generic form-creator, or a "page" for every modelType?
  - What's the status on UI-forms? Can we start integrating this soon?

**Comment by Mohammer5** [comment]
Thanks @Birkbjo for going through the code and sorry for waiting for so long to reply

## Structure

> Import paths for selectors might be a bit annoying though.

I've experimented a little with re-exporting index files in my private projects and I start to really like it.
So if we have a `src/redux/index.js` that re-exports everything that's used by the app (action creators, selectors & `configureStore`), import paths will be ok again (`import { getFooBar } from '../../redux'`)

> I do however think that 'components' should be grouped into 'feature'-folders. They kind of are now, but I would move files like 'ProtectedRoute.js' and 'Navigation.js' into 'Navigation' folder.
> I would move each file in the root of the "components"-folders into a sub-folder, some of the folders currently only have .css, I would expect these to live with the .js files.

I think that's a very good suggestions. I'll prepare a PR

> I'm unsure about what should go in the 'pages'-folder. We are probably going to have generic components like 'ModelTypeList' and 'ModelForm'. Are these going in the page-folder or component-folder? Or is this just to setup routes for "overrides" (I know, that word has bad reputation by now).

I'd put all components into pages that are specified as component on the `<Route />`s. I view them as controller code (in an MVC sense), so they don't contain logic themselves but compose the route's functionality from features entirely

## Other

> We need an 'initial'-loader. This needs to load userSettings, systemSettings, userAuthorities, userDataStore. This should probably be stored in redux.

This is an interestnig part. As of now I think @varl and @amcgee want us to use the data provider for such data; and once cached, we can use it in different places throughout the app (@amcgee another reason why we need caching soon).

> This also means that we can remove the queries for Navigation in useHasAuthority, as these can be checked synchronously from the redux-store after initialization.

If my assumption above is true, these would have to stay, so this depends on the decision how/where to store/fetch data

> Right now the NavigationLinkComponent will send a request for every rendered link . . . I still believe having one bigger request on load for this is better than using the Data-Engine to send requests for each link.

Yeah you're probably right. I'd like to do this once caching is available, otherwise we'd load the whole data the whole time.

> I don't think we need 'connected-react-router'. Use history and push from there if we need to do this imperatively.

Yes! I don't know why I didn't try that, thanks for pointing that out

> We should use Link-components instead of onClick-handlers calling 'push', unless there's a good reason not to.

This depends on how we implement `ui-core` components. Right now I'm using the `MenuItem` and it doesn't support using a custom component

> I think sectionOrder can be simplified by moving the options-object to the section-objects. I was initially confused by the optional array-with-option API.

:100: 

## Way forward

> 'view-data-transformation' (might be done with selectors?).

Yes, both redux and the data engine (in the future? @amcgee is there a roadmap / already-included-features list?) support selectors

> Did we decide on a generic form-creator, or a "page" for every modelType?

We haven't decided yet. My preferred way would be to have a page for each modelType (controller code style) and then compose it from features. This will most likely result in duplicate code, but I think in this case that's a good case because 1. it's easy to add exceptions and 2. it decreases complexity quite a lot

> What's the status on UI-forms? Can we start integrating this soon?

I'd venture to say: yes
The basic fields are all more or less ready (select is almost ready, it's being finalized right now), the Org Unit Tree is being finalized right now as well. The Transfer component hasn't been started yet (previously: GroupEditor component)

<hr />

**Off topic**: After having worked with the data engine a bit, I think I'd really like to have to functionality in redux available. It's awesome in react for simple things and recursive loading logic (org unit tree). For flat and complicated/shared api data, it's introducing complexity in the architecture as we'd have to separate loading data (through moving data up the tree & either passing down the data or using context) from displaying it, which is the reason why flux/redux got so popular in the first place

<hr />

@dhis2/team-apps What are your thoughts on the issues raised by @Birkbjo and my response?

**Comment by ismay** [comment]
> I've experimented a little with re-exporting index files in my private projects and I start to really like it. So if we have a src/redux/index.js that re-exports everything that's used by the app (action creators, selectors & configureStore), import paths will be ok again (import { getFooBar } from '../../redux')

I like that too. I'll add that I also like using default exports if it's the only thing exported from a file (see https://github.com/airbnb/javascript/issues/1365#issuecomment-293052808 for a discussion on the why).

**Comment by HendrikThePendric** [comment]
> I'm unsure about what should go in the 'pages'-folder. We are probably going to have generic components like 'ModelTypeList' and 'ModelForm'. Are these going in the page-folder or component-folder? Or is this just to setup routes for "overrides" (I know, that word has bad reputation by now).

The way I imagined this to work is:
- Rename `pages` to `routes`, because I think that's a better fit conceptually
- For some sections these will be fully custom
- And I think the generic list and form should be included there too, i.e. `ModelList` and `ModelForm`

> Decide on how much schema-information we want to utilize, and if so; we may need to load these on initial-load as well

IMO the schema information becomes valuable when editing a specific model, as such I don't think any of this should be included in the initial loading phase, but fetched when accessing a create/edit form...

Saying that, the schemas do also contain useful information such as displayNames. Depending on how the schemas works reg. translations it could possibly be useful to fetch the displayNames for all sections when the app boots. We should look into that.

> so it might be nice to have this in one place, instead of the 'feature/duck'-way.

I was quite looking forward to trying the "duck" approach in the new Maintenance App, but no big deal...

> > Did we decide on a generic form-creator, or a "page" for every modelType?

> We haven't decided yet. My preferred way would be to have a page for each modelType (controller code style) and then compose it from features. This will most likely result in duplicate code, but I think in this case that's a good case because 1. it's easy to add exceptions and 2. it decreases complexity quite a lot

I feel like we keep re-opening this discussion, even though we decided on this a fairly long time ago. Perhaps I am wrong, but in my mind the reasoning was as follows:
1. I prepared [a breakdown](https://docs.google.com/spreadsheets/d/1kCotR9SCyHtT8F4aTtrfhJs72qTbNtQRS0omT9sJR4w/edit#gid=0) of all individual sections of the maintenance app
2. In this breakdown we can see that approx. 2/3 of the sections follow a common pattern.
3. As such, some form of generic List and Form components seems to be a very logical direction. A little bit of duplicate code isn't a bad thing, but not for 2/3 of the sections.

Perhaps I've misunderstood what @Mohammer5 meant. Just having a page/route that returns the generic Form/List with some props would be fine. But really, that's just a semantic issue, either keeping the configuration in a POJO or as configuring via component props....

> What's the status on UI-forms? Can we start integrating this soon?

Unfortunately, it has halted and won't be worked on until the `next` branch of `ui-core` has gone into master. There are going to be quite a few breaking changes on the form elements in ui-core.

**Comment by Mohammer5** [comment]
> Rename pages to routes, because I think that's a better fit conceptually

I disagree with renaming the pages to sections:
* The router maps routes to pages, it resolves the route and responds with a page
* I'd expect a route to look like this:
```js
<Route
  path="...."
  ....
/>
```

I'm not against renaming `pages`, but I think `routes` is wrong

> And I think the generic list and form should be included there too, i.e. ModelList and ModelForm

> Just having a page/route that returns the generic Form/List with some props would be fine. 

That's exactly what I mean. The 2/3 more or less identical sections will all have their own route/page/resolver, but ideally they do nothing but rendering a `ModelList` or `ModelForm` with some custom props (e. g. like `mode="organisationUnit"`; just an example, no idea if that's what we'll need to pass):

```js
// src/pages/organisationUnit/edit.js
export const Edit = props => (
  <ModelForm id={props.id} model="organisationUnit" />
)
```

This will reduce the amount of config files and resolving dynamic stuff, etc. But if there's a lot of common stuff, these pages shouldn't contain more than 5-10 lines of code or so

**Comment by HendrikThePendric** [comment]
> I'm not against renaming pages, but I think routes is wrong

I don't think either is right or wrong per se. Strictly speaking, we are not rendering a route (as @Mohammer5 explains), nor are we rendering a page. Perhaps `view` is fairly apt and it fits in with the MVC convention mentioned earlier. In any case, I'm not too bothered about this part....

> That's exactly what I mean.

OK, great. Then that's fine.

**Comment by Mohammer5** [comment]
@HendrikThePendric Oh yeah, `view` is good! It also aligns with how phone apps call their pages/routes/views!

> I was quite looking forward to trying the "duck" approach in the new Maintenance App, but no big deal...

We'll still have this, just for the few redux states (navigation, userSettings, systemSettings, userAuthorities, userDataStore, ...), all the global stuff. I think redux can be considered its own feature aka global state. But as it includes setup components on the root level (the `Provider`) and has state, I'd like to keep it a separate folder. Everything else will be in the `duck` way.

I think you've proposed this style before and I wasn't sure what to think about it, because I could see potential maintenance issues as technologies (like redux) would then be all over the place. I've played around with it a bit in private projects and think in the end, it doesn't really matter what we do as long as we keep it as simple as possible. Then it'll be maintainable no matter what we do, so grouping features might actually be nice.

**Comment by Mohammer5** [comment]
I think all issues address here have been solved!

@Birkbjo's issues:
- [x] move each file in the root of the "components"-folders into a sub-folder
  Done
- [x] Import paths for selectors might be a bit annoying though
  Each folder on the root level now has an `index.js` exporting the folder's contents
- [x] We need an 'initial'-loader. This needs to load userSettings, systemSettings, userAuthorities, userDataStore. This should probably be stored in redux
  Now in redux store and loaded initially. There's a `Loading` component without styles.
- [x] Right now the NavigationLinkComponent will send a request for every rendered link. This is mostly Data-Engines fault, but can be improved by the above point.
  See item above
- [x] I don't think we need 'connected-react-router'
  Removed
- [x] I think sectionOrder can be simplified by moving the options-object to the section-objects
  Simplified

@HendrikThePendric's issues:
- [x] Rename `pages` to `views`
  Done
